### ISSUE TITLE
TimeDistributed returns a different tensor shape in tf.keras and keras.

### ISSUE THREAD

**Issue by kiflowb777** [keras-team/keras#13497]
**System information**  
- Have I written custom code (as opposed to using example directory):  
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.14-gpu
- Keras version:  2.2.4 and 2.2.4-tf
- Python version:  3.6.1
- CUDA/cuDNN version:  10/6.7.4
- GPU model and memory:  RTX 2060 6GB

You can obtain the TensorFlow version with:  
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"   unknown 1.14.0
You can obtain the Keras version with:  
python -c 'import keras as k; print(k.__version__)'  

When trying to use the fpn_classifier_graph from Mask_RCNN I get different tensor shapes depending on keras version (https://github.com/matterport/Mask_RCNN/issues/1820).

In the internal function: `x = KL.TimeDistributed(KL.BatchNormalization(), name='mrcnn_class_bn1')(x, training=train_bn`
the shape of x is:
`1. Tensor("mrcnn_class_bn1/Reshape_1:0", shape=(?, 1000, 1, 1, 256), dtype=float32)`
`2. Tensor("mrcnn_class_bn1/Reshape_1:0", shape=(?, ?, 1, 1, 256), dtype=float32)`
(1.) is Keras, (2.) is tf.keras.

I also write simple code to check TimeDistributed behavior in keras and tf.keras:
```
use_tf_keras = False

if use_tf_keras:
    from tensorflow.keras.layers import Dense
    from tensorflow.keras.layers import TimeDistributed
    from tensorflow.keras.backend import constant
else:
    from keras.layers import Dense
    from keras.layers import TimeDistributed
    from keras.backend import constant
import numpy as np


a = np.array([[[1,2,3],[1,2,3],[1,2,3]]])
x = constant(a)
x = TimeDistributed(Dense(8), input_shape=(10, 16))(x)
x
```
and by changing the value of the variable use_tf_keras (True, False) I get:
`<tf.Tensor 'time_distributed_5/Reshape_1:0' shape=(1, 3, 8) dtype=float32>`
`<tf.Tensor 'time_distributed_36/transpose_1:0' shape=(1, ?, 8) dtype=float32>`

**Comment by kiflowb777** [comment]
The reason why layers have different shapes is because the original layer used in Mask RCNN is a custom layer which does not call the compute_output_shape if we use tf.keras.

More info you can find here: https://github.com/tensorflow/tensorflow/issues/33785

**Comment by ibrahimLearning** [comment]
> 
> 
> The reason why layers have different shapes is because the original layer used in Mask RCNN is a custom layer which does not call the compute_output_shape if we use tf.keras.
> 
> More info you can find here: [tensorflow/tensorflow#33785](https://github.com/tensorflow/tensorflow/issues/33785)

@kiflowb777 
So How can I solve it? Would you give some suggestions please. Thanks

**Comment by kiflowb777** [comment]
@ibrahimLearning
In each layer inheriting from the class KE.Layer in Mask RCNN code (e.g. class ProposalLayer(KE.Layer))
you have to add call: `self.compute_output_shape(self, input_shape)` at the end of `def call(self, inputs):` method.

Add also tensorflow before each keras import: (e.g.` import keras; from keras... -> import tensorflow.keras; from tensorflow.keras`)  
https://github.com/tensorflow/tensorflow/issues/33786

And fix model.metrics_tensors:
 `model.metrics_tensors.append(loss) -> model.metrics.append(loss) `
https://github.com/tensorflow/tensorflow/issues/33789

Try also tf-nightly-gpu and watch: https://github.com/tensorflow/tensorflow/pull/34248

**Comment by kartik4949** [comment]
> @ibrahimLearning
> In each layer inheriting from the class KE.Layer in Mask RCNN code (e.g. class ProposalLayer(KE.Layer))
> you have to add call: `self.compute_output_shape(self, input_shape)` at the end of `def call(self, inputs):` method.
> 
> Add also tensorflow before each keras import: (e.g.` import keras; from keras... -> import tensorflow.keras; from tensorflow.keras`)
> [tensorflow/tensorflow#33786](https://github.com/tensorflow/tensorflow/issues/33786)
> 
> And fix model.metrics_tensors:
> `model.metrics_tensors.append(loss) -> model.metrics.append(loss) `
> [tensorflow/tensorflow#33789](https://github.com/tensorflow/tensorflow/issues/33789)
> 
> Try also tf-nightly-gpu and watch: [tensorflow/tensorflow#34248](https://github.com/tensorflow/tensorflow/pull/34248)

self.compute_output_shape(self,input_shape) is it a valid call??

**Comment by kiflowb777** [comment]
To fix this try:

```
def call(self, input):
   input_shape = input.shape  # you must find a way to retrieve this value on your own
   ...
   out_shape = self.compute_output_shape(input_shape)
   output = output.set_shape(out_shape)
   return output
```

More info: https://github.com/tensorflow/tensorflow/issues/33785
